src/espcomp/mod.rs

```diff
@@ -0,0 +1,129 @@
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+use flate2::read::GzDecoder;
+use log::info;
+use tar::Archive;
+
+pub mod comp;
+
+const API_BASE_URL: &str = "https://api.components.espressif.com";
+
+struct ApiClient {
+    base_url: String,
+}
+
+impl ApiClient {
+    pub fn new() -> Self {
+        Self {
+            base_url: API_BASE_URL.to_string(),
+        }
+    }
+
+    pub fn get_component(&self, namespace: &str, name: &str) -> Result<comp::WithVersions> {
+        let url = format!("{}/components/{namespace}/{name}", self.base_url);
+        let component = ureq::get(&url).call()?.into_json::<comp::WithVersions>()?;
+        Ok(component)
+    }
+}
+
+pub struct EspComponentDep {
+    pub namespace: String,
+    pub name: String,
+    pub version_req: semver::VersionReq,
+}
+
+impl EspComponentDep {
+    pub fn new(namespace: String, name: String, version_req: semver::VersionReq) -> Self {
+        Self { namespace, name, version_req }
+    }
+}
+
+pub struct EspComponentManager {
+    managed_component_dir: PathBuf,
+    components: Vec<EspComponentDep>,
+    api_client: ApiClient,
+}
+
+impl EspComponentManager {
+    pub fn new(component_dir: PathBuf) -> Self {
+        Self { managed_component_dir: component_dir, components: vec![], api_client: ApiClient::new() }
```

Comment:
Nit: maybe just rename `managed_component_dir` to just `component_dir`?

Reply:
Renamed to `components_dir`, I think the pluralization makes sense here.